Enable MPU for SAM D5x and E5x

stephanosio opened the pull request:
This PR enables the ARM memory protection unit for the Atmel SAM D5x and E5x series SoCs (SAM D51, SAM E51, SAM E53 and SAM E54).

This PR also includes miscellaneous commits that fix the related issues.

Tested and verified working on `atsame54_xpro`.

~Includes the commits from #24721 (blocked by #24721).~ Merged
~Includes the commits from #24740 (blocked by #24740).~ Merged

stephanosio committed:
soc: arm: same54: Add MPU selection

This commit adds the missing `CPU_HAS_ARM_MPU` selection for the Atmel
SAM E54 series SoC.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
dts: arm: atmel: Add MPU node for SAM D5x and E5x

The Atmel SAM D5x and E5x series SoCs include the ARM memory protection
unit that supports up to 8 memory regions.

This commit adds the missing device tree MPU node to `samd5x.dtsi`,
which is used by both the SAM D5x and E5x series SoCs.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: samd51: Add MPU selection

This commit adds the missing `CPU_HAS_ARM_MPU` selection for the Atmel
SAM D51 series SoC.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: same51: Add MPU selection

This commit adds the missing `CPU_HAS_ARM_MPU` selection for the Atmel
SAM E51 series SoC.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: same53: Add MPU selection

This commit adds the missing `CPU_HAS_ARM_MPU` selection for the Atmel
SAM E53 series SoC.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: arm: atsame54_xpro: Enable MPU

This commit enables the ARM memory protection unit for the SAM D54
Xplained Pro board, since `CPU_HAS_ARM_MPU` is now selected by the SoC
Kconfig.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
boards: arm: atsame54_xpro: Use emulation reset for OpenOCD

The OpenOCD JTAG reset causes the SAM E54 SoC to prematurely start code
execution before the SoC is ready, and this causes hard faults when the
MPU is enabled -- causing OpenOCD to hang.

This commit disables the JTAG SRST reset mechanism and falls back to
using the Cortex-M SYSRESETREQ emulation reset mechanism, which does
not have experience this issue.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio commented:
Maybe it's the CI, or general lack of interest in this platform; but, I keep digging up more broken builds -- now I2C and SPI ...

ioannisg commented on dts/arm/atmel/samd5x.dtsi:
```diff
@@ -17,6 +17,14 @@
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			mpu: mpu@e000ed90 {
```
That's redundant, for now, but no harm.

ioannisg commented on include/arch/arm/aarch32/cortex_m/nvic.h:
```diff
@@ -7,6 +7,8 @@
 #ifndef ZEPHYR_INCLUDE_ARCH_ARM_AARCH32_CORTEX_M_NVIC_H_
 #define ZEPHYR_INCLUDE_ARCH_ARM_AARCH32_CORTEX_M_NVIC_H_
 
+#include <devicetree.h>
```
Why do we need to include devicetree.h here? Where is it explicitly required?

stephanosio replied:
If you take a look a few lines below:
```
#define NVIC_NODEID DT_INST(0, arm_v8m_nvic)
#elif defined(CONFIG_ARMV7_M_ARMV8_M_MAINLINE)
#define NVIC_NODEID DT_INST(0, arm_v7m_nvic)
#elif defined(CONFIG_ARMV6_M_ARMV8_M_BASELINE)
#define NVIC_NODEID DT_INST(0, arm_v6m_nvic)
#endif
#define NUM_IRQ_PRIO_BITS DT_PROP(NVIC_NODEID, arm_num_irq_priority_bits)
```

ioannisg replied:
I will need to understand why this inclusion is required now, and not when this header was introduced.

stephanosio replied:
Many SoCs tend to include `devicetree.h` somewhere along their headers, but that is apparently not the case for SAM E5x.

Either way, as long as the header makes use of device tree macros which are defined by `devicetree.h`, it should be including the `devicetree.h`.

ioannisg replied:
> Many SoCs tend to include devicetree.h somewhere along their headers, but that is apparently not the case for SAM E5x.

Then I would insist on having SAM E5x SoCs include devicetree.h in soc.h. We started introducing this convention a while back; most of ARM SoCs already have transitioned to include devicetree.h in their soc.h. We should do it for the few ARM SoCs that are remaining.

(Having this kind of consistent behavior across ARM SoCs helps very much in identifying inclusion cycles - something we have troubled a lot with - in ARM builds and makes it easier to do reviews of SoC definitions across different ARM vendors)

Regarding nvic.h: this header was introduced very recently, and is included in just two places, cmsis.h and exc.h and _both_ should be including devicetree.h before including nvic.h, so you don't need to include devicetree.h here (although not an error or a bad practice, this would help finding inconsistencies in the code, such as the mising devicetree inclusion in E5x socs.).

stephanosio replied:
> Regarding nvic.h: this header was introduced very recently, and is included in just two places, cmsis.h and exc.h and _both_ should be including devicetree.h before including nvic.h, so you don't need to include devicetree.h here

For `exc.h`, it includes `devicetree.h` before including `nvic.h`, so no problem there.
For `cmsis.h`, it includes `soc.h` before including `nvic.h`, which means `soc.h` (or one of the headers included by it) must include `devicetree.h`.

My question is: why have such a convoluted and indirect scheme when we can just `#include <devicetree.h>` in `nvic.h` as one would normally do? It seems `devicetree.h` only includes the headers that have one-way dependency and should never experience cross dependency issues.

stephanosio replied:
> Then I would insist on having SAM E5x SoCs include devicetree.h in soc.h. We started introducing this convention a while back; most of ARM SoCs already have transitioned to include devicetree.h in their soc.h. We should do it for the few ARM SoCs that are remaining.

I see that only about half of the SoCs seem to be following this convention. I do not necessarily object to this convention; but is this still relevant? (i.e. do we actually have a problem if we don't do this?) As far as I can see, `devicetree.h` is very well isolated and does not have any cross dependencies.

ioannisg replied:
@stephanosio in master branch, if you do not include devicetree.h in soc.h you cannot compile for user-mode. You can try this on E54.

This matter was identified 18 months back, after we eliminated a huge sort of inclusion cycles. Then we decided that by convention, all soc.h (at least for ARM) has to include devicetree.h (At that time, was called, generated_dts_board.h).

The SoCs that don't follow the convention yet are likely the ones that do not have MPU at all, or they don' have it yet enabled, such as the ATSAME54.

Now:

A week ago we added the nvic.h header. The reason was mostly to have NUM_PRIO_BITS included in assembly code. This header is included in cmsis.h which already soc.h, which already includes devicetree.h (or it has to, according to our convention). nvic.h header was never meant to change any inclusion paths! (its content could well be copy-pasted in the two places the header is included, or it could just be an internal-hidden header,. just to not duplicate the added content).

Including devicetree.h in nvic.h is not desired (but not wrong of course), because it hides the problem with SoCs not following the convention to include devicetree in the soc header.

stephanosio replied:
> in master branch, if you do not include devicetree.h in soc.h you cannot compile for user-mode. You can try this on E54.

I can successfully build and run `tests/kernel/mem_protect/userspace` and other userspace-enabled tests and samples on E54, which does not have `#include <devicetree.h>` in `soc.h` though.

stephanosio replied:
Maybe the recent `devicetree.h` rework has something to do with it? At least, for now, it no longer seems including `devicetree.h` in `soc.h` is a requirement.

stephanosio replied:
For now, I have added `devicetree.h` inclusion in all Atmel SAM(0) family SoCs, until we figure out if there is a better way to handle this after 2.3 release.

ioannisg reviewed (commented):
Looks mostly ok. I suppose it should be rebased after the port macro in the samples, tests is renamed.
I've got just one concern about including devicetree.h

stephanosio committed:
soc: arm: atmel_sam0: Add devicetree.h inclusion in soc.h

This commit adds the `devicetree.h` header inclusion in the Atmel SAM0-
family SoC header files, as required by the ARM SoC conventions.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

stephanosio committed:
soc: arm: atmel_sam: Add devicetree.h inclusion in soc.h

This commit adds the `devicetree.h` header inclusion in the Atmel SAM-
family SoC header files, as required by the ARM SoC conventions.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>

tyrantlizard77 reviewed (approved):
LGTM. I defer to @galak and @ioannisg when it comes to devicetree.

nandojve commented on soc/arm/atmel_sam/sam3x/soc.h:
```diff
@@ -17,6 +17,9 @@
 
 #ifndef _ASMLANGUAGE
 
+/* Add include for DTS generated information */
+#include <devicetree.h>
+
```
same here.

stephanosio replied:
https://github.com/zephyrproject-rtos/zephyr/pull/24715#discussion_r415945019

stephanosio commented on soc/arm/atmel_sam0/samd20/soc.h:
```diff
@@ -13,6 +13,9 @@
 #include <zephyr/types.h>
 
+/* Add include for DTS generated information */
+#include <devicetree.h>
+
```
https://github.com/zephyrproject-rtos/zephyr/pull/24715#discussion_r415945019

stephanosio replied:
Also the commit message:
```
soc: arm: atmel_sam: Add devicetree.h inclusion in soc.h

This commit adds the `devicetree.h` header inclusion in the Atmel SAM-
family SoC header files, as required by the ARM SoC conventions.

Signed-off-by: Stephanos Ioannidis <root@stephanos.io>
```

stephanosio commented:
Rebased

stephanosio commented:
All the dependencies have been merged. This PR is ready for merging.
